### PR TITLE
fix: use offsets in ListRef::value_at()

### LINKED ISSUE #4833: bug: wrong return value of array access
**Describe the bug**
Get unexpected output.

**To Reproduce**
To create table and insert values
```
CREATE TABLE bid (
    "price" BIGINT,
    "channel" VARCHAR,
    "url" VARCHAR
);

INSERT INTO bid (
    price,
    channel,
    url
) VALUES
(73134520, 'channel-7568', 'query=1&channel_id=163053568'),
(1940, 'channel-9319', 'query=1&channel_id=433848320'),
(12655, 'channel-5136', 'query=1&channel_id=136839168'), 
(19269, 'channel-5775', 'query=1&channel_id=244842496'), 
(43672280, 'channel-9550', 'query=1&channel_id=1923350528'), 
(28953332, 'channel-5698', 'query=1&channel_id=1114112000');
```
query
```
SELECT
    price, channel,
    (regexp_match(url, '(&|^)channel_id=([^&]*)'))[1]
    AS channel_id
FROM
    bid Order by price;
```

**Expected behavior**
Risingwave's output:
```
  price   |   channel    | channel_id
----------+--------------+------------
     1940 | channel-9319 | 433848320
    12655 | channel-5136 | 136839168
    19269 | channel-5775 | 1923350528
 28953332 | channel-5698 | 136839168
 43672280 | channel-9550 | 1923350528
 73134520 | channel-7568 | 163053568
```

The output for the same input in PostgreSQL
```
1940	channel-9319	433848320
12655	channel-5136	136839168
19269	channel-5775	244842496
28953332	channel-5698	1114112000
43672280	channel-9550	1923350528
73134520	channel-7568	163053568
```
**Additional context**
The problem was found while adding nexmark21 support.
Any help is welcomed.

### PR DESCRIPTION
**Pull request by nanderstabel**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Fix `value_at()` method in `ListRef`

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- use `self.offsets[idx]` in `ListRef::value_at(idx)` in order to return the correct value. Previously in some instances the wrong value was returned because always offset == 0 was used by default.
- add several unit tests for ListRef and filter executor and min() list.
- add test_minmax from DuckDB.
- add frontend test for ArrayAccess

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

Documentation
If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

Types of user-facing changes
Please keep the types that apply to your changes, and remove those that do not apply.

* SQL commands, functions, and operators

Release note
Unquoted array style syntax (`{1, 2, 3}`) is now deprecated.
Single quoted array style syntax (`'{1, 2, 3}'`) is now implemented.

## related issue

close #4833

### CODE REVIEW

File: e2e_test/batch/aggregate/duckdb/test_minmax.slt.part [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# Copied from https://github.com/duckdb/duckdb (MIT licensed).
+# Copyright 2018-2022 Stichting DuckDB Foundation
+
+statement ok
+CREATE TABLE lists(l int[]);
+
+statement ok
+INSERT INTO lists VALUES ({0, 10}), ({1, 11}), ({2, 12}), ({3, 13}), ({4, 14}), ({5, 15});
```
Comment (neverchanje) [inline e2e_test/batch/aggregate/duckdb/test_minmax.slt.part:8]: I fail this query on postgres. I guess it's a duckdb-specific syntax?

Comment (neverchanje) [reply e2e_test/batch/aggregate/duckdb/test_minmax.slt.part:8]:
```sql
INSERT INTO lists VALUES ('{0, 10}'), ('{1, 11}'), ('{2, 12}'), ('{3, 13}'), ('{4, 14}'), ('{5, 15}');
```

This syntax works. The duckdb one will break our compatibility with postgres, though I do prefer this syntax.

Comment (wzzzzd) [reply e2e_test/batch/aggregate/duckdb/test_minmax.slt.part:8]: risingwave supports creating array using 'array[1, 2, 3]' or '{1, 2, 3}' and pg does not support the latter...

Comment (nanderstabel) [reply e2e_test/batch/aggregate/duckdb/test_minmax.slt.part:8]: Ok yes indeed the syntax should be compatible with pg, will fix.

**Review by xxchan — COMMENTED** [review]
Could you also add array access e2e tests in `array.slt.part`?

e.g., for the bug found in this issue https://github.com/singularity-data/risingwave/issues/4833

```
dev=> create table t(x int[]);
CREATE_TABLE
dev=> insert into t values (Array[0,1]),(Array[0,2]),(Array[0,3]), (Array[0,4]),(Array[0,5]);
INSERT 0 5
dev=> select x[2] from t;
 ?column?
----------
        3
        1
        2
        4
        4
(5 rows)
```

**Comment by nanderstabel** [comment]
> Could you also add array access e2e tests in `array.slt.part`?
> 
> e.g., for the bug found in this issue #4833
> 
> ```
> dev=> create table t(x int[]);
> CREATE_TABLE
> dev=> insert into t values (Array[0,1]),(Array[0,2]),(Array[0,3]), (Array[0,4]),(Array[0,5]);
> INSERT 0 5
> dev=> select x[2] from t;
>  ?column?
> ----------
>         3
>         1
>         2
>         4
>         4
> (5 rows)
> ```

Will do 👍

File: e2e_test/batch/aggregate/duckdb/test_minmax.slt.part (hunk 1 above)
Comment (nanderstabel) [reply e2e_test/batch/aggregate/duckdb/test_minmax.slt.part]: > ```sql
> INSERT INTO lists VALUES ('{0, 10}'), ('{1, 11}'), ('{2, 12}'), ('{3, 13}'), ('{4, 14}'), ('{5, 15}');
> ```
> 
> This syntax works. The duckdb one will break our compatibility with postgres, though I do prefer this syntax. The postgres way is like deferring the parsing phase to string casting.

Solution for accepting single-quoted '{1, 2, 3}' arrays (like pg does) as opposed to {1, 2, 3} like we currently do. 

Three scenario's:
- `ARRAY[1]` should be parsed as `integer[]`: completed
- `{1}` syntax without single quotes: is now deprecated in this PR
- `'{1}'` pg-like syntax: is now implemented in this PR

The problem with `'{1}'` style syntax is that it is ambiguous for the parser: it is a valid `Varchar` value as well as a valid `integer[]` value. 

That's why `'{1}'` will be parsed as a SingleQuoteString and bound to a `Literal::new(Some(ScalarImpl::Utf8(s)), DataType::Varchar)`. Depending on the context this value can now be implicitly/assign/explicitly casted to a `List`. 

Most relevant e2e tests can be found in `e2e_test/batch/basic/array.slt.part` under `## Test single quote style array syntax.`.

**Review by fuyufjh — APPROVED** [review]

**Pull request merged**